### ISSUE TITLE
Force stack to use local packages over Stackage?

### ISSUE THREAD

**Issue by sid-kap** [commercialhaskell/stack#2181]
I'm trying to build a package using, as a dependency, the current version of a project in the dependency's Github repo. I've added that repo's Git location and commit ID to the `packages` section of my `stack.yaml`. 

The problem is that I think that Stack might be selecting the version in the Stackage lts snapshot instead of the version that I specified in `stack.yaml`. (I'm not sure. Is there a way to check?)

_If_ this is indeed happening, I think it might be happening because the dependency's version number on Stackage is higher than the version number in the git repo (even though the version in the git repo is actually newer, and is the one I want to build against). (To be concrete, the dependency has version 1.3.1.2 on Stackage, but the cabal file of the version in the repo says version 1.3.) 

So, in summary:

1. Is there a way for me to check which versions of the dependencies Stack is building against (and where it's getting the dependencies from)?
2. How exactly does Stack choose how to satisfy its dependencies? Does the `extra-dep` flag change this behavior?
3. If it is the case that Stack is choosing the lts version over the Git version that I specified, then is there a way for me to tell Stack which version to use?

**Comment by harendra-kumar** [comment]
* Use `list-dependencies` command to list all the dependencies
* Yes the `extra-dep` package flag will make this package satisfy the dependencies in your project instead of using a snapshot version. It is just like as if you specified it in the extra-deps section. See the documentation for more details on the flag syntax and behavior - http://docs.haskellstack.org/en/stable/yaml_configuration/.

**Comment by mgsloan** [comment]
`stack exec -- ghc-pkg describe pkg` is also useful to find out what versions something was built with.

**Comment by sid-kap** [comment]
`stack exec -- ghc-pkg describe pkg` shows me both versions (the Git version and the Stackage version):
```
name: diagrams-lib
version: 1.3
id: diagrams-lib-1.3-c0229e120dbed411f10ebeddf37ace7a
key: diagr_6pqMzBz1lpPKMET5uLg8cg
license: BSD3
maintainer: diagrams-discuss@googlegroups.com
homepage: http://projects.haskell.org/diagrams
synopsis: Embedded domain-specific language for declarative graphics
description:
    Diagrams is a flexible, extensible EDSL for creating
    graphics of many types.  Graphics can be created
    in arbitrary vector spaces and rendered with
    multiple backends.  diagrams-lib provides a
    standard library of primitives and operations for
    creating diagrams.  To get started using it, see
    the "Diagrams" module, and refer to the tutorials and
    documentation on the diagrams website,
    <http://projects.haskell.org/diagrams>.
category: Graphics
author: Brent Yorgey
exposed: True
exposed-modules:
    Diagrams Diagrams.Prelude Diagrams.Align Diagrams.Angle
    Diagrams.Animation Diagrams.Animation.Active Diagrams.Attributes
    Diagrams.Attributes.Compile Diagrams.Backend.CmdLine
    Diagrams.BoundingBox Diagrams.Combinators Diagrams.Coordinates
    Diagrams.CubicSpline Diagrams.CubicSpline.Boehm
    Diagrams.CubicSpline.Internal Diagrams.Deform Diagrams.Direction
    Diagrams.Envelope Diagrams.LinearMap Diagrams.Located
    Diagrams.Names Diagrams.Parametric Diagrams.Parametric.Adjust
    Diagrams.Path Diagrams.Points Diagrams.Query Diagrams.Segment
    Diagrams.Size Diagrams.Tangent Diagrams.ThreeD
    Diagrams.ThreeD.Align Diagrams.ThreeD.Attributes
    Diagrams.ThreeD.Camera Diagrams.ThreeD.Deform Diagrams.ThreeD.Light
    Diagrams.ThreeD.Shapes Diagrams.ThreeD.Size
    Diagrams.ThreeD.Transform Diagrams.ThreeD.Projection
    Diagrams.ThreeD.Types Diagrams.ThreeD.Vector Diagrams.Trace
    Diagrams.Trail Diagrams.TrailLike Diagrams.Transform
    Diagrams.Transform.ScaleInv Diagrams.Transform.Matrix Diagrams.TwoD
    Diagrams.TwoD.Adjust Diagrams.TwoD.Align Diagrams.TwoD.Arc
    Diagrams.TwoD.Arrow Diagrams.TwoD.Arrowheads
    Diagrams.TwoD.Attributes Diagrams.TwoD.Combinators
    Diagrams.TwoD.Curvature Diagrams.TwoD.Deform Diagrams.TwoD.Ellipse
    Diagrams.TwoD.Image Diagrams.TwoD.Model Diagrams.TwoD.Offset
    Diagrams.TwoD.Path Diagrams.TwoD.Points Diagrams.TwoD.Polygons
    Diagrams.TwoD.Segment Diagrams.TwoD.Segment.Bernstein
    Diagrams.TwoD.Size Diagrams.TwoD.Shapes Diagrams.TwoD.Text
    Diagrams.TwoD.Transform Diagrams.TwoD.Types Diagrams.TwoD.Vector
    Diagrams.Util
trusted: False
import-dirs: /home/sidharth/clones/diagrams-doc/.stack-work/install/x86_64-linux/lts-5.17/7.10.3/lib/x86_64-linux-ghc-7.10.3/diagrams-lib-1.3-6pqMzBz1lpPKMET5uLg8cg
library-dirs: /home/sidharth/clones/diagrams-doc/.stack-work/install/x86_64-linux/lts-5.17/7.10.3/lib/x86_64-linux-ghc-7.10.3/diagrams-lib-1.3-6pqMzBz1lpPKMET5uLg8cg
data-dir: /home/sidharth/clones/diagrams-doc/.stack-work/install/x86_64-linux/lts-5.17/7.10.3/share/x86_64-linux-ghc-7.10.3/diagrams-lib-1.3
hs-libraries: HSdiagrams-lib-1.3-6pqMzBz1lpPKMET5uLg8cg
depends:
    JuicyPixels-3.2.7-ea64a73602300c0ed0b1acd16698272b
    active-0.2.0.9-b988306edc150f07c9bfc2b582f554ee
    adjunctions-4.3-55f60f723ea689fec8b5c34540e658f6
    array-0.5.1.0-960bf9ae8875cc30355e086f8853a049
    base-4.8.2.0-0d6d1084fbc041e1cded9228e80e264d
    cereal-0.5.1.0-b8a30f8877fa63f4f84ceceaab55715e
    colour-2.3.3-0c4e3735159b357562a94a22fd309019
    containers-0.5.6.2-59326c33e30ec8f6afd574cbac625bbb
    data-default-class-0.0.1-fcbab95490d730dd2489b628499587cd
    diagrams-core-1.3.0.4-4c330dafa48a17f4353292d490af9cfc
    diagrams-solve-0.1.0.1-90bf91390e3170024e0324330a8d31a8
    directory-1.2.2.0-660a7a83a753ed85c8a374c15dae2b97
    distributive-0.5.0.2-393a023f71dc3cd11f1fef5933233cdf
    dual-tree-0.2.0.9-21cfca261100f373ef1194a58201f6f4
    exceptions-0.8.2.1-08ed62d2282590eb80eba3e5c42a5f4e
    filepath-1.4.0.0-f97d1e4aebfd7a03be6980454fe31d6e
    fingertree-0.1.1.0-40f1e6845ddd53fd219763dc87c9ce79
    fsnotify-0.2.1-43d29731a8a2c0468d554b3b87bf184d
    hashable-1.2.4.0-eb8410705fa05638eebff0131e5dac80
    intervals-0.7.2-1e0886725cc18a761c086c5caa552da5
    lens-4.13-33e36cce992ef417de5ab19125cf35a7
    linear-1.20.4-527d6edd97c9de38f17909eceadc26cc
    monoid-extras-0.4.0.2-f1a99edd38e31a96f27577bc31f05fd5
    mtl-2.2.1-3af90341e75ee52dfc4e3143b4e5d219
    optparse-applicative-0.12.1.0-e571c94d6c730b4cb5b36eba77024bb6
    process-1.2.3.0-a22328103774f0499a990653944cbf99
    profunctors-5.2-73596694dd7b7d9c2ee52c51c668d7ea
    semigroups-0.18.1-dcd63a094a9869a49e49ccebf8ed01a2
    tagged-0.8.4-9a04b6ba5af4681c23b22a393e2724e7
    text-1.2.2.1-f5b1651133426aa2c83c19d64d248779
    transformers-0.4.2.0-81450cd8f86b36eaa8fa0cbaf6efc3a3
    unordered-containers-0.2.5.1-efcb998195e61e370da9dd9c1bcddcff
haddock-interfaces: /home/sidharth/clones/diagrams-doc/.stack-work/install/x86_64-linux/lts-5.17/7.10.3/doc/diagrams-lib-1.3/diagrams-lib.haddock
haddock-html: /home/sidharth/clones/diagrams-doc/.stack-work/install/x86_64-linux/lts-5.17/7.10.3/doc/diagrams-lib-1.3
pkgroot: "/home/sidharth/clones/diagrams-doc/.stack-work/install/x86_64-linux/lts-5.17/7.10.3"
---
name: diagrams-lib
version: 1.3.1.2
id: diagrams-lib-1.3.1.2-924e679439cbbcefe3e82f68cd04dcd5
key: diagr_ENIESF25GBcHLMK9mNfYfq
license: BSD3
maintainer: diagrams-discuss@googlegroups.com
homepage: http://projects.haskell.org/diagrams
synopsis: Embedded domain-specific language for declarative graphics
description:
    Diagrams is a flexible, extensible EDSL for creating
    graphics of many types.  Graphics can be created
    in arbitrary vector spaces and rendered with
    multiple backends.  diagrams-lib provides a
    standard library of primitives and operations for
    creating diagrams.  To get started using it, see
    the "Diagrams" module, and refer to the tutorials and
    documentation on the diagrams website,
    <http://projects.haskell.org/diagrams>.
category: Graphics
author: Brent Yorgey
exposed: True
exposed-modules:
    Diagrams Diagrams.Prelude Diagrams.Align Diagrams.Angle
    Diagrams.Animation Diagrams.Animation.Active Diagrams.Attributes
    Diagrams.Attributes.Compile Diagrams.Backend.CmdLine
    Diagrams.BoundingBox Diagrams.Combinators Diagrams.Coordinates
    Diagrams.CubicSpline Diagrams.CubicSpline.Internal Diagrams.Deform
    Diagrams.Direction Diagrams.Envelope Diagrams.LinearMap
    Diagrams.Located Diagrams.Names Diagrams.Parametric
    Diagrams.Parametric.Adjust Diagrams.Path Diagrams.Points
    Diagrams.Query Diagrams.Segment Diagrams.Size Diagrams.Tangent
    Diagrams.ThreeD Diagrams.ThreeD.Align Diagrams.ThreeD.Attributes
    Diagrams.ThreeD.Camera Diagrams.ThreeD.Deform Diagrams.ThreeD.Light
    Diagrams.ThreeD.Shapes Diagrams.ThreeD.Size
    Diagrams.ThreeD.Transform Diagrams.ThreeD.Projection
    Diagrams.ThreeD.Types Diagrams.ThreeD.Vector Diagrams.Trace
    Diagrams.Trail Diagrams.TrailLike Diagrams.Transform
    Diagrams.Transform.ScaleInv Diagrams.Transform.Matrix Diagrams.TwoD
    Diagrams.TwoD.Adjust Diagrams.TwoD.Align Diagrams.TwoD.Arc
    Diagrams.TwoD.Arrow Diagrams.TwoD.Arrowheads
    Diagrams.TwoD.Attributes Diagrams.TwoD.Combinators
    Diagrams.TwoD.Curvature Diagrams.TwoD.Deform Diagrams.TwoD.Ellipse
    Diagrams.TwoD.Image Diagrams.TwoD.Model Diagrams.TwoD.Offset
    Diagrams.TwoD.Path Diagrams.TwoD.Points Diagrams.TwoD.Polygons
    Diagrams.TwoD.Segment Diagrams.TwoD.Segment.Bernstein
    Diagrams.TwoD.Size Diagrams.TwoD.Shapes Diagrams.TwoD.Text
    Diagrams.TwoD.Transform Diagrams.TwoD.Types Diagrams.TwoD.Vector
    Diagrams.Util
trusted: False
import-dirs: /home/sidharth/.stack/snapshots/x86_64-linux/lts-5.17/7.10.3/lib/x86_64-linux-ghc-7.10.3/diagrams-lib-1.3.1.2-ENIESF25GBcHLMK9mNfYfq
library-dirs: /home/sidharth/.stack/snapshots/x86_64-linux/lts-5.17/7.10.3/lib/x86_64-linux-ghc-7.10.3/diagrams-lib-1.3.1.2-ENIESF25GBcHLMK9mNfYfq
data-dir: /home/sidharth/.stack/snapshots/x86_64-linux/lts-5.17/7.10.3/share/x86_64-linux-ghc-7.10.3/diagrams-lib-1.3.1.2
hs-libraries: HSdiagrams-lib-1.3.1.2-ENIESF25GBcHLMK9mNfYfq
depends:
    JuicyPixels-3.2.7-ea64a73602300c0ed0b1acd16698272b
    active-0.2.0.9-af351d8076bd627a9824ff6dd82597ae
    adjunctions-4.3-55f60f723ea689fec8b5c34540e658f6
    array-0.5.1.0-960bf9ae8875cc30355e086f8853a049
    base-4.8.2.0-0d6d1084fbc041e1cded9228e80e264d
    colour-2.3.3-0c4e3735159b357562a94a22fd309019
    containers-0.5.6.2-59326c33e30ec8f6afd574cbac625bbb
    data-default-class-0.0.1-fcbab95490d730dd2489b628499587cd
    diagrams-core-1.3.0.7-78daff2dd35d9070b46a50994aad2a11
    diagrams-solve-0.1.0.1-7d6a5a907abc27302db9386f35fe5e49
    directory-1.2.2.0-660a7a83a753ed85c8a374c15dae2b97
    distributive-0.5.0.2-393a023f71dc3cd11f1fef5933233cdf
    dual-tree-0.2.0.9-6c3144c858902169b9df45e7a0818e28
    exceptions-0.8.2.1-08ed62d2282590eb80eba3e5c42a5f4e
    filepath-1.4.0.0-f97d1e4aebfd7a03be6980454fe31d6e
    fingertree-0.1.1.0-40f1e6845ddd53fd219763dc87c9ce79
    fsnotify-0.2.1-43d29731a8a2c0468d554b3b87bf184d
    hashable-1.2.4.0-eb8410705fa05638eebff0131e5dac80
    intervals-0.7.2-1e0886725cc18a761c086c5caa552da5
    lens-4.13-33e36cce992ef417de5ab19125cf35a7
    linear-1.20.4-527d6edd97c9de38f17909eceadc26cc
    monoid-extras-0.4.0.4-1d04958b713dd1b44c3e5fa35d09fb72
    mtl-2.2.1-3af90341e75ee52dfc4e3143b4e5d219
    optparse-applicative-0.12.1.0-e571c94d6c730b4cb5b36eba77024bb6
    process-1.2.3.0-a22328103774f0499a990653944cbf99
    semigroups-0.18.1-dcd63a094a9869a49e49ccebf8ed01a2
    tagged-0.8.4-9a04b6ba5af4681c23b22a393e2724e7
    text-1.2.2.1-f5b1651133426aa2c83c19d64d248779
    transformers-0.4.2.0-81450cd8f86b36eaa8fa0cbaf6efc3a3
    unordered-containers-0.2.5.1-efcb998195e61e370da9dd9c1bcddcff
haddock-interfaces: /home/sidharth/.stack/snapshots/x86_64-linux/lts-5.17/7.10.3/doc/diagrams-lib-1.3.1.2/diagrams-lib.haddock
haddock-html: /home/sidharth/.stack/snapshots/x86_64-linux/lts-5.17/7.10.3/doc/diagrams-lib-1.3.1.2
pkgroot: "/home/sidharth/.stack/snapshots/x86_64-linux/lts-5.17/7.10.3"
```

**Comment by sid-kap** [comment]
I think I found the answer to my last question: Each version in the output of `stack exec -- ghc-pkg describe dependency-name` has a unique hash, and so if I do `stack exec -- ghc-pkg describe my-package` I see the hash of the one that is used.

**Comment by sid-kap** [comment]
Before I close this, I still would like to know the answers to my questions 2 and 3 above. I will play around with this for a few minutes and report my observations.

**Comment by sid-kap** [comment]
Okay, it seems like Stack _is_ doing what I wanted it to do, so there was actually no problem here. It was building against the local/Git version that I specified, rather than the Stackage version.

It would be nice to document this behavior. Is there a formal spec somewhere where we say how Stack chooses dependencies? If not, can we start one?

**Comment by mgsloan** [comment]
Not really a formal spec, but: http://docs.haskellstack.org/en/stable/architecture/ and http://docs.haskellstack.org/en/stable/GUIDE/ .  Do those not address this point?

I would prefer to hold off on writing down anything too elaborate until https://github.com/commercialhaskell/stack/issues/1265 is implemented.  Not that the semantics will change too drastically, but if it mentions details of implementation, that may change.

**Comment by sid-kap** [comment]
As far as I can tell, GUIDE doesn't have anything relevant to this. The "Shadowing" section of the architecture page (http://docs.haskellstack.org/en/stable/architecture/#shadowing), however, explains exactly what I wanted to know. 

I think we should put something about this in the [Yaml Configuration page, under the "Project config" subheader](http://docs.haskellstack.org/en/stable/yaml_configuration/#project-config). In particular, under the "packages", "complex package locations", "Local dependency packages", and "extra-deps" sections, I would like to add something like:

> Note: If you specify a package here, it shadows any package of the same name from the snapshot resolver. See [the Shadowing section of the architecture page](http://docs.haskellstack.org/en/stable/yaml_configuration/#project-config) for more details.

Alternatively, we could just add _one_ note under the "resolver" heading (also under "Yaml Configuration/Project config"), like:

> Note: Any packages specified under `packages` or `extra-deps` "shadows" all packages of the same name from the resolver. See [the Shadowing section of the architecture page](http://docs.haskellstack.org/en/stable/yaml_configuration/#project-config) for more details.

I believe this information belongs in Yaml configuration page (or at least some other prominent location, like the GUIDE) because it is part of the semantics of the Yaml file. In the section that explains what to put yaml file, I believe we ought to specify how those configurations are interpreted by Stack (the semantics).

**Comment by sid-kap** [comment]
Does this sound reasonable? If so, I can submit a PR with these changes to the doc.

**Comment by harendra-kumar** [comment]
Sounds good to me. It will explicitly state the behavior which I found missing when I checked it after you asked this question. We can add a note in "Local Dependency packages" because it talks about the `extra-dep` package flag. We can also add the note in the resolver section as you suggested.

**Comment by mgsloan** [comment]
Sounds good to me too, thanks!

**Comment by sid-kap** [comment]
@harendra-kumar Why only in the "Local Dependency packages", and not in the "packages", "complex package locations", and "extra-deps" sections? The situation is the same in all four cases. I think we should either mention this in all four sections, or in none of them.

Ideally, we should find one place to keep it, instead of repeating ourselves 5 times.

**Comment by sid-kap** [comment]
I'm going to draft a short paragraph to place directly under "Package config". We can discuss this once I open the PR.

Thanks @harendra-kumar and @mgsloan for all your help!
